### PR TITLE
Change Mapping Rules reuse to a durable function

### LINKED ISSUE #688: Optimise rule generation
In #675 we moved to serverless generating Mapping Rules. 
But further work is necessary to support (very) large scan reports, and prepare for greater scale overall.

A loose plan is to move towards orchestrating the serverless functions, with [Durable Functions](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-overview?tabs=in-process%2Cnodejs-v3%2Cv1-model&pivots=python#publications).

There is also:
- additional optimisation to happen in the rules algorithm itself
- technical debt in the existing functions that still use the Django API rather than connect to the DB directly.

Some of this work is exploratory - as durable functions is new to the team. 

# Tasks
- [x] New (http?) durable function "Orchestrating" `CreateConcepts` and `MappingRules`
- [x] Change the `CreateConcepts` + `MappingRules` to be able to be orchestrated
- [x] Change the triggering from Django form to point to the new Orchestrator
- [x] Change `refresh_mapping_rules` make deleting existing rules optional.
- [ ] Memoize the destination tables in `_save_mapping_rules`
- [ ] Remove unnecessary lookups in `_save_mapping_rules`
- [x] Explore whether the "dequeuing" is still relevant in durable functions

## Lower Priority
- [x] Explore "status" endpoint of durable functions, is this easy for us to implement as a quick way to give the user a status of the mapping? To make the UX less "eerie"

# Acceptance Criteria:
- [x] Reusing mapping rules is durable for scale
- [x] Reduce complexity of the mapping rules algo
- [x] Tests pass

### PR DESCRIPTION
**Pull request by AndyRae**
# Changes

Updates the reuse of Mapping Rules to used Durable Azure Functions. Enables reusing rules to be much quicker, and more reliable.

Includes:
- Introduces a http trigger for the rules to be reused
- Updates the `CreateConcepts` and `MappingRules` functions to be activity functions.
- Updates the Django API to call the trigger when the table ID and date event are selected, instead of queuing a job. 
- The Django app is now responsible for deleting the existing rules.
- The trigger runs the orchestrator function, which: 
    - Calls `CreateConcepts` function
    - "Fans out" to parallelise the `MappingRules` function, based on paginating the created concepts.
- Delete the `DatabaseTest` and `ProcessQueue` functions

The durable functions also include job statuses, which we can use in the future to monitor the status of the reuse, and make it less "eery". 

I made the decision to not update the algorithm for saving rules, for time, and to persist with smaller changes.

Closes #688 

# Environment changes needed

Add the following environment variables to the Django app:
- [ ] AZ_FUNCTIONS_URL
- [ ] AZ_FUNCTIONS_RULES_NAME

# Checks

**Important:** please complete these **before** merging.
- [x] Run migrations, if any.
- [x] Update `changelog.md`, including migration instructions if any.
- [x] Run unit tests.

### CODE REVIEW

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/shared/shared/services/rules.py [hunk 1]
```diff
@@ -85,6 +90,35 @@ def _find_existing_concepts(table_id: int) -> List[ScanReportConcept]:
...
+    Args:
+        - table_id (int): Id of the ScanReportTable to filter by.
+
+    Returns:
+        - The count of ScanReportConcepts attached to the Table Id.
+    """
+    # Count the number of ScanReportConcepts associated with the table_id
+    concept_count = (
+        ScanReportValue.objects.filter(scan_report_field__scan_report_table=table_id)
+        .filter(concepts__isnull=False)
```
[the remark this answers is not in the document]
Comment (AndyRae) [reply app/shared/shared/services/rules.py:106]: You mean the below one (Just to check?)
That would be great - but I can't think of a way to do it. This is a reworking of the `find_existing_concepts`

[the remark this answers is not in the document]
Comment (dcl10) [reply app/shared/shared/services/rules.py:106]: I mean won't `.filter(scan_report_field__scan_report_table=table_id, concepts__isnull=False)` work?

File: app/api/api/views.py [hunk 2]
```diff
@@ -567,6 +569,9 @@ def partial_update(self, request: Any, *args: Any, **kwargs: Any) -> Response:
         serializer.is_valid(raise_exception=True)
         self.perform_update(serializer)
 
+        # Delete the current mapping rules
+        delete_mapping_rules(instance.id)
+
```
[the remark this answers is not in the document]
Comment (AndyRae) [reply app/api/api/views.py:574]: `instance = self.get_object()`
will raise a 404 if instance does not exist.

[the remark this answers is not in the document]
Comment (dcl10) [reply app/api/api/views.py:574]: Ah okay cool

File: app/workers/CreateConcepts/__init__.py [hunk 3]
```diff
@@ -322,10 +323,4 @@ def main(msg: func.QueueMessage):
 
     asyncio.run(_handle_table(table, vocab_dictionary))
 
-    # send to MappingRules queue
-    message = {
-        "table_id": table_id,
-    }
-    add_message(os.environ.get("MAPPING_RULES_QUEUE_NAME"), message)
-
     return
```
[the remark this answers is not in the document]
Comment (AndyRae) [reply app/workers/CreateConcepts/__init__.py]: Thanks - have removed

File: app/shared/shared/services/rules.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AndyRae) [reply app/shared/shared/services/rules.py:106]: Sorry I misunderstood! I thought you meant combine the two `concept_count` db hits, but yes you're right

File: app/api/api/views.py [hunk 4]
```diff
@@ -575,13 +581,22 @@ def partial_update(self, request: Any, *args: Any, **kwargs: Any) -> Response:
...
+            response = requests.post(orchestrator_url, json=msg)
+            response.raise_for_status()
+        except request.exceptions.HTTPError as e:
+            logging.error(f"HTTP Trigger failed: {e}")
+
+        # TODO: The worker_id can be used for status, but we need to save it somewhere.
+        # resp_json = response.json()
+        # worker_id = resp_json.get("id")
 
         return Response(serializer.data)
```
Comment (dcl10) [inline app/api/api/views.py:601]: Should the returned `Response` object be different and have another code depending on whether the `POST` succeeded or not?

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/api/api/views.py (hunk 4 above)
Comment (AndyRae) [reply app/api/api/views.py:601]: Short answer - ideally yes, if we fail to send to the worker, then the user should know about it. But we have nothing on the frontend to handle that anyway. 

So in our current Carrot setup no, this is the existing behaviour - you click the button, and hope it works. If it breaks, you'll never know... 
Which is related to the commented out bit - now we have a job ID that we can check for status also.

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/shared/shared/services/rules.py [hunk 5]
```diff
@@ -49,23 +49,28 @@ def _delete_mapping_rules(table_id: int) -> None:
...
 
     Returns:
         - A list of ScanReportConcept attached to the Table Id.
     """
+    offset = page * page_size
 
     values = (
         ScanReportValue.objects.all()
         .filter(scan_report_field__scan_report_table=table_id)
         .filter(concepts__isnull=False)
```
[the remark this answers is not in the document]
Comment (AndyRae) [reply app/shared/shared/services/rules.py]: Yes but that's not my changes ! :D I have fixed this

**Review by dcl10 — APPROVED** [review]

**Pull request merged**